### ISSUE TITLE
#pragma unroll with extra assumptions compiles to sub optimal assembly

### ISSUE THREAD

**Issue by MkazemAkhgary** [ispc/ispc#2476]
```c
export void sum(uniform float a[], uniform float b[], uniform float c[], uniform int num) {
    assume(num%programCount==0);
    // this assumption should help to produce better code
    assume(num>=programCount*2);
    #pragma unroll 2
    foreach(i = 0 ... num)
    {
        c[i] = a[i] + b[i];
    }
}
```

for --target=avx2-i32x8, This compiles to

```asm
sum:                                    # @sum
        mov     r9d, 8
        mov     r8d, 32
.LBB2_1:                                # =>This Inner Loop Header: Depth=1
        lea     eax, [r8 - 32]
        cdqe
        vmovups ymm0, ymmword ptr [rdi + rax]
        vaddps  ymm0, ymm0, ymmword ptr [rsi + rax]
        vmovups ymmword ptr [rdx + rax], ymm0
        cmp     r9d, ecx    # ******* not needed
        jge     .LBB2_3      # ******* not needed
        movsxd  rax, r8d
        vmovups ymm0, ymmword ptr [rdi + rax]
        vaddps  ymm0, ymm0, ymmword ptr [rsi + rax]
        vmovups ymmword ptr [rdx + rax], ymm0
        lea     eax, [r9 + 16]
        add     r8d, 64
        add     r9d, 8
        cmp     r9d, ecx
        mov     r9d, eax
        jl      .LBB2_1
.LBB2_3:
        vzeroupper
        ret
```

The assumption of "num>=programCount*2" can be used to avoid cmp and jge instruction.  note that with the same assumption, if we wanted to unroll 4, number of checks required can still be halved.

**Comment by MkazemAkhgary** [comment]
There could also be some room for instruction scheduling, although I'm not sure how complicated it would be for more general cases. but in this simple example, instead of reusing ymm0 register, ymm1 could be used to execute second vaddps. that should help with instruction pipelining, I haven't done benchmarks though. Just wanted to share my thoughts.

**Comment by MkazemAkhgary** [comment]
I have noticed that this issue has been resolved since ispc 1.21 therefore I'm closing this.

```asm
sum: # @sum
        dec     ecx
        shr     ecx, 3
        inc     ecx
        mov     r9d, ecx
        and     r9d, -2
        xor     r8d, r8d
.LBB1_1:                                # =>This Inner Loop Header: Depth=1
        mov     eax, r8d
        and     eax, -64
        vmovups ymm0, ymmword ptr [rdi + rax]
        vaddps  ymm0, ymm0, ymmword ptr [rsi + rax]
        vmovups ymmword ptr [rdx + rax], ymm0
        vmovups ymm0, ymmword ptr [rdi + rax + 32]
        vaddps  ymm0, ymm0, ymmword ptr [rsi + rax + 32]
        vmovups ymmword ptr [rdx + rax + 32], ymm0
        add     r8, 64
        add     r9d, -2
        jne     .LBB1_1
        test    cl, 1
        je      .LBB1_4
        and     r8d, -64
        vmovups ymm0, ymmword ptr [rdi + r8]
        vaddps  ymm0, ymm0, ymmword ptr [rsi + r8]
        vmovups ymmword ptr [rdx + r8], ymm0
.LBB1_4:
        vzeroupper
        ret
```
